### ISSUE TITLE
unstable step inside gitlabCommitStatus marks stage as success in gitlab

### ISSUE THREAD

**Issue by cedricje** [jenkinsci/gitlab-plugin#713]
### Context
- **Gitlab plugin version**: 1.5.3
- **Gitlab version**: GitLab Community Edition 10.4.2
- **Jenkins version**: 2.104
- **Job type**: Pipeline

### Logs & Traces
For reproduction i've made a minimal pipeline script with just one gitlabCommitStatus, and a fixed junit xml with a failure, which marks the build unstable.

```groovy
    gitlabBuilds(builds: ['stage2']) {
            stage('stage2') {
                gitlabCommitStatus(name: 'stage2') {
                    sh 'touch results/*.xml'
                    junit 'results/*.xml'
                   echo "Result: $currentBuild.result"
            }
        }
    }
```
[results_xml.zip](https://github.com/jenkinsci/gitlab-plugin/files/1719303/results_xml.zip)

### Problem description

_Describe your problem in a meaningful way_:
- a merge request triggers a build in Jenkins. 
 the build has a gitlabCommitStatus block with a junit step which marks the build as unstable in Jenkins.
 When printing currentBuild.result inside the gitlabCommitStatus block, this also prints UNSTABLE
- i expected this stage to be marked as failed in gitlab. From #190 i'd expect this is the default behavior.
- the actual result is the pipeline stage is marked as passed in gitlab.
  
Did the default to mark unstable build as failed change? And how can i set this option?

**Comment by omehegan** [comment]
Somewhat related to #386.

**Comment by omehegan** [comment]
@cedricje for one thing, check and make sure that the act of running the `echo` step does not then flip the status of that block from unstable to successful. Can you also check and see if a failing junit step does cause Jenkins to send a failure status to GitLab? And can you check that with and without the `echo` step you're using now?

**Comment by omehegan** [comment]
@cedricje ping?

**Comment by cedricje** [comment]
@omehegan : sorry, did the tests and forgot to post the results. I couldn't get builds to fail with JUnit, i redid the tests with xUnit instead

- same result after removing the `echo` step
- a Jenkins failure isn't sent as a failure status to Gitlab. I did the test with and without `echo` step.

**Comment by tamasgal** [comment]
The same happens for me. The build is marked as unstable in Jenkins but every stage is green in Gitlab. For me also the JUnit test is the one which marks the build as unstable, when one or more tests are failing.

Is there a workaround or update for this?

**Comment by omehegan** [comment]
@tamasgal can you share the Pipeline code you are using in this case? Also, if this is a regular Pipeline job, is 'Mark unstable builds as success' checked or unchecked?

**Comment by tamasgal** [comment]
Uhm, I am now using XUnit, which I configured to mark the build as failed, instead of unstable...

**Comment by tamasgal** [comment]
This was the Jenkinsfile as far as I remember: http://git.km3net.de/km3py/km3pipe/blob/239866f2bf63bc9aea2fdcf693507ddea2d82288/Jenkinsfile

**Comment by omehegan** [comment]
@tamasgal so now what you are seeing is that when XUnit marks the job as failed, the status in GitLab still shows as successful? Can you look at the Pipeline Steps log for a job that shows this issue, and see which step is failing? You should see the build status being sent at that point. If you follow the instructions in the User Support section of the README to increase debug logging for the plugin, you should see log entries (in the main Jenkins log, not the build log) showing what status is being sent.

**Comment by apallier** [comment]
Hello, I have the same problem with this pipeline code:

```
        stage('Test') {
            steps {
            gitlabCommitStatus("Test") {
                sh '''
                    pytest ${PYTEST_OPTIONS} ${COV_OPTIONS} ./src/ ./scripts/ || :
                '''
                junit testResults: 'tests_result.xml'
                cobertura coberturaReportFile: 'coverage.xml', onlyStable: false
            }}
        }
```

In this case, one test failed by status in GitLab is "Success"

**Comment by apallier** [comment]
I found this workaround but don't understand why it works.
Is Jenkins 'UNSTABLE' status reported as 'failed' in GitLab or not? 

```
stage('Test') {
    steps {
        gitlabCommitStatus("Test") {
            sh '''
                pytest ${PYTEST_OPTIONS} ${COV_OPTIONS} ./src/framework/titanframework || :
            '''
            junit testResults: 'tests_result.xml'
            cobertura coberturaReportFile: 'coverage.xml', onlyStable: false
        }
        script {
            if (currentBuild.result == 'UNSTABLE') {
                updateGitlabCommitStatus name: "Test", state: 'failed'
            }
        }
    }
}
```

**Comment by omehegan** [comment]
@apallier that's very interesting. In Freestyle jobs, there is an option when you publish build status, to mark 'unstable' as 'success.' It is disabled by default. But I wonder if it ends up enabled in a Pipeline job, because there is no UI to set it. I'll have to experiment.

**Comment by KyleWiering** [comment]
Adding a new thing that behaves differently then before and removes previous functionality is a major change - I didn't see the major version bump. Is this plugin using semantic versioning? (even if a toggle was added to recover the previous functionality - it is still a major change)

**Comment by omehegan** [comment]
@KyleWiering what new thing are you referring to that was added?

**Comment by omehegan** [comment]
@apallier I can reproduce this. Freestyle builds with unstable result send a failed status to GitLab, but Pipeline builds seem to send a success status. I will look into this further.

**Comment by Symbianx** [comment]
@omehegan we have the same problem, is there a workaround to set the status in gitlab as failed and maintain the jenkins build as unstable?

**Comment by omehegan** [comment]
@Symbianx yes, see https://github.com/jenkinsci/gitlab-plugin/issues/713#issuecomment-412875379

**Comment by ababushk** [comment]
Happily, junit step sets build result to UNSTABLE not when build ends, but right after it finishes uploading xml files, so we're using something like that (but it looks ugly):
```
gitlabCommitStatus(name: 'e2e_tests') {
  sh "./run_tests.sh"
  junit "*/TEST-*.xml"
  if (currentBuild.result == "UNSTABLE") {
    error("Some tests not passed, fail the build")
   }
}
```
Also, `currentBuild.resultWorseOrEqualTo('UNSTABLE')` can be used

**Comment by Hawk777** [comment]
I have experienced two problems with the workarounds noted so far.

apallier’s workaround, for me anyway, (1) results in the “Test” pipeline stage showing up twice in Gitlab, as if there were two separate stages of the same name, the first successful and the second failed (the first from the `gitlabCommitStatus` block and the second from the `updateGitlabCommitStatus` call); and (2) because it only checks the build status at that point, would result in marking that stage as failed if the build were unstable for a reason stemming from earlier in the pipeline, not only if something within that block caused it to become unstable. The latter could be worked around by checking whether the status was `SUCCESS` before starting the block and changed to `UNSTABLE` afterwards, but that comes with its own problems: if the status was `UNSTABLE` before starting the block but became “more unstable” within the block (some more failed tests were reported), the stage really ought to be reported as failed. I’m not sure how easy this would be to implement, as I know nothing about the Jenkins API and how granularly it reports changes in build status. But ideally, if Gitlab presents information as if each stage were independent, it would be nice for exactly those stages that contribute to the unstableness to be reported as failing, and not the others.

oxygenxo’s workaround has the same problem (2), and also AFAICT (I haven’t tested) also sets the Jenkins status to `FAILURE` when it would otherwise be `UNSTABLE`, which is not ideal (I would prefer that a build that otherwise worked be unstable, not failed).

I found a workaround in which I check the exit code from my test program and, after recording its output with `junit`, throw an exception if tests failed. The exception passing through the `gitlabCommitStatus` block causes Gitlab to show failure. I catch the exception (and check its message for a magic string) and swallow it outside the `gitlabCommitStatus` block to prevent it from propagating up to the top level and changing the Jenkins status from `UNSTABLE` to `FAILURE`. This seems to solve all problems: the Gitlab status is only shown once, it’s exactly the outcome of this stage of tests that produces that status, and the Jenkins status remains `UNSTABLE` when I want it to be. But it relies on being able to detect failed tests by a means other than the `junit` step, and the code is really ugly. Basically it looks like this:

```
script {
  try {
    gitlabCommitStatus(name: 'Tests 1') {
      script {
        int exitCode = sh(script: './run-tests.sh', returnStatus: true);
        junit('test-results.xml');
        if(exitCode == 1) {
          throw new Exception('GITLAB_STATUS_WORKAROUND');
        }
      }
    }
  } catch(Exception exp) {
    if(exp.getMessage().equals('GITLAB_STATUS_WORKAROUND')) {
      // Swallow exception.
    } else {
      throw exp;
    }
  }
}
```

**Comment by Hawk777** [comment]
It looks like there has been a recent API change involving stages reporting build statuses. [Jenkins-39203](https://issues.jenkins-ci.org/browse/JENKINS-39203) and [Jenkins-45579](https://issues.jenkins-ci.org/browse/JENKINS-45579) seem relevant. Is that something that this plugin could use to identify where instability comes from in the interest of reporting it accurately?

**Comment by commodis** [comment]
Using the workaround works for me. No need to try catch or use other plugins

```groovy
steps {
    gitlabCommitStatus('step-name') {
        warnError('commands which can fail to produce an UNSTABLE build result') {
            sh "exit 1"
        }

        script {
            if (currentBuild.result == 'UNSTABLE') {
                updateGitlabCommitStatus name: 'step-name', state: 'failed'
            }
        }
    }
}
```
